Cubical/Experiments/List.agda

```diff
@@ -0,0 +1,204 @@
+
+  rev-++-distr' : (xs ys : List' A) → rev' (xs ++' ys) ≡ rev' ys ++' rev' xs
+  rev-++-distr' {A = A} = transport (λ i → (xs ys : ListPath A i)
+                                         → revP i (appP i xs ys) ≡ appP i (revP i ys) (revP i xs))
+                                    rev-++-distr
+    where
+    appP : PathP (λ i → ListPath A i → ListPath A i → ListPath A i) _++_ _++'_
+    appP i = transp (λ j → ListPath A (i ∧ j) → ListPath A (i ∧ j) → ListPath A (i ∧ j)) (~ i) _++_
+
+    revP : PathP (λ i → ListPath A i → ListPath A i) rev rev'
+    revP i = transp (λ j → ListPath A (i ∧ j) → ListPath A (i ∧ j)) (~ i) rev
+
+
+
+-- The above operations for List' are derived by going back and
+-- forth. With the SIP we can do better and transport properties for
+-- user defined operations (assuming that the operations are
+-- well-defined wrt to the forward direction of the equivalence).
+open import Cubical.Foundations.SIP
+open import Cubical.Structures.Axioms
+open import Cubical.Structures.Auto
+
+module SIP (A : Type) where
+
+  -- First define the raw structure without axioms. This is just the
+  -- signature of _++_ and rev.
+  RawStruct : Type → Type
+  RawStruct X = (X → X → X) × (X → X)
+
+  -- Some automated SIP magic
+  RawEquivStr : _
+  RawEquivStr = AutoEquivStr RawStruct
+
+  rawUnivalentStr : UnivalentStr _ RawEquivStr
+  rawUnivalentStr = autoUnivalentStr RawStruct
+
+  -- Now the property that we want to transport
+  P : (X : Type) → RawStruct X → Type
+  P X (_++_ , rev) = ((xs ys : X) → rev (xs ++ ys) ≡ rev ys ++ rev xs)
+
+  -- Package things up for List
+  List-Struct : Σ[ X ∈ Type ] (Σ[ s ∈ RawStruct X ] (P X s))
+  List-Struct = List A , (_++_ , rev) , rev-++-distr
+
+
+  -- We now give direct definitions of ++' and rev' for List'
+  _++'_ : List' A → List' A → List' A
+  [] ++' ys        = ys
+  (xs ∷' x) ++' ys = (xs ++' ys) ∷' x
+
+  rev' : List' A → List' A
+  rev' [] = []
+  rev' (xs ∷' x) = rev' xs ++' ([] ∷' x)
+
+  -- We then package this up into a raw structure on List'
+  List'-RawStruct : Σ[ X ∈ Type ] (RawStruct X)
+  List'-RawStruct = List' A , (_++'_ , rev')
+
+  -- Finally we prove that toList' commutes with _++_ and rev. Note
+  -- that this could be a lot more complex, see for example the Matrix
+  -- example (Cubical.Algebra.Matrix).
+  toList'-++ : (xs ys : List A) → toList' (xs ++ ys) ≡ toList' xs ++' toList' ys
+  toList'-++ [] ys = refl
+  toList'-++ (x ∷ xs) ys i = toList'-++ xs ys i ∷' x
+
+  toList'-rev : (xs : List A) → toList' (rev xs) ≡ rev' (toList' xs)
+  toList'-rev [] = refl
+  toList'-rev (x ∷ xs) = toList'-++ (rev xs) (x ∷ []) ∙ cong (_++' ([] ∷' x)) (toList'-rev xs)
+
+  -- We can now get the property for ++' and rev' via the SIP
+  goal : (xs ys : List' A) → rev' (xs ++' ys) ≡ rev' ys ++' rev' xs
+  goal = transferAxioms rawUnivalentStr List-Struct List'-RawStruct
+                        (ListEquiv , toList'-++ , toList'-rev) .snd .snd
```

Comment:
Maybe `transferAxioms` should really just output the axioms? We can then have `inducedStructure` instead of the current `transferAxioms`?

Reply:
I did this now. What do you think? @cangiuli @ecavallo @mzeuner